File: frontend/src/services/attendanceService.js
import api from './api';

const attendanceService = {
    // Get attendance by date
    getAttendanceByDate: async (date) => {
        const response = await api.get(`/attendance/date/${date}`);
        return response.data;
    },

    // Get attendance by student
    getAttendanceByStudent: async (studentId, date = null) => {
        const url = date ? `/attendance/student/${studentId}?date=${date}` : `/attendance/student/${studentId}`;
        const response = await api.get(url);
        return response.data;
    },

    // Get attendance by subject
    getAttendanceBySubject: async (subjectId) => {
        const response = await api.get(`/attendance/subject/${subjectId}`);
        return response.data;
    },

    // Mark single attendance
    markAttendance: async (attendanceData) => {
        const response = await api.post('/attendance/mark', attendanceData);
        return response.data;
    },

    // Mark self attendance
    markSelfAttendance: async (date) => {
        const response = await api.post('/attendance/student/mark', { date });
        return response.data;
    },

    // Mark bulk attendance
    markBulkAttendance: async (attendanceList) => {
        const response = await api.post('/attendance/bulk', attendanceList);
        return response.data;
    },

    // Update attendance
    updateAttendance: async (id, attendanceData) => {
        const response = await api.put(`/attendance/${id}`, attendanceData);
        return response.data;
    },

    // Get attendance report
    getAttendanceReport: async (studentId, startDate, endDate) => {
        const response = await api.get(`/attendance/report/${studentId}`, {
            params: { startDate, endDate }
        });
        return response.data;
    },

    // Get attendance percentage
    getAttendancePercentage: async (studentId, subjectId) => {
        const response = await api.get(`/attendance/percentage/${studentId}/${subjectId}`);
        return response.data;
    },

    // Get today's attendance
    getTodayAttendance: async () => {
        const response = await api.get('/attendance/today');
        return response.data;
    }
};

export default attendanceService;

File: backend/src/main/java/com/erp/attendance/AttendanceService.java
package com.erp.attendance;

import com.erp.attendance.dto.AttendanceRequest;

import java.util.List;
import java.util.Map;

public interface AttendanceService {
    Attendance markAttendance(AttendanceRequest request);
    List<Attendance> markBulkAttendance(List<AttendanceRequest> requests);
    Attendance markSelfAttendance(Long studentId, java.time.LocalDate date);
    List<Attendance> getAttendanceByStudent(Long studentId);
    List<Attendance> getAttendanceByStudentAndDate(Long studentId, java.time.LocalDate date);
    List<Attendance> getAttendanceBySubject(Long subjectId);
    List<Attendance> getAttendanceByDate(java.time.LocalDate date);
    Attendance updateAttendance(Long id, AttendanceRequest request);
    Map<String, Object> getAttendanceReport(Long studentId);
    Map<String, Object> getAttendanceReport(Long studentId, java.time.LocalDate startDate, java.time.LocalDate endDate);
    Double getAttendancePercentage(Long studentId, Long subjectId);
    Map<String, Object> getClassAttendanceReport(Long teacherId, Long subjectId, String date);
}
